File: js/src/utf8.mjs
// UTF-8 codec, hand-written on purpose.
//
// The host's own facilities are not used: TextDecoder is lenient (it substitutes
// U+FFFD where the spec demands E_UTF8), and JS string indexing counts UTF-16
// units, which disagrees with PHP on every code point above U+FFFF. Every length,
// offset and slice in SEL counts code points, and that has to be true in both
// hosts or nothing else is.

import { fail } from './errors.mjs';

// --- code points ------------------------------------------------------------

// A JS string with an unpaired surrogate has no UTF-8 encoding, so it cannot be
// a SEL TEXT value.
export function toCodePoints(str, pos) {
  const out = [];
  for (let i = 0; i < str.length; i++) {
    const c = str.charCodeAt(i);
    if (c >= 0xd800 && c <= 0xdbff) {
      const d = i + 1 < str.length ? str.charCodeAt(i + 1) : 0;
      if (d < 0xdc00 || d > 0xdfff) fail('E_UTF8', 'unpaired high surrogate', pos);
      out.push(0x10000 + ((c - 0xd800) << 10) + (d - 0xdc00));
      i++;
    } else if (c >= 0xdc00 && c <= 0xdfff) {
      fail('E_UTF8', 'unpaired low surrogate', pos);
    } else {
      out.push(c);
    }
  }
  return out;
}

export function fromCodePoints(cps) {
  let out = '';
  // Chunked to stay clear of argument-count limits on long strings.
  for (let i = 0; i < cps.length; i += 4096) {
    out += String.fromCodePoint.apply(null, cps.slice(i, i + 4096));
  }
  return out;
}

export function cpLength(str, pos) {
  return toCodePoints(str, pos).length;
}

// start is 0-based here; the 1-based convention belongs to the built-ins.
export function cpSlice(str, start, len, pos) {
  const cps = toCodePoints(str, pos);
  return fromCodePoints(cps.slice(start, len === undefined ? undefined : start + len));
}

// --- bytes ------------------------------------------------------------------

export function encodeUtf8(str, pos) {
  const cps = toCodePoints(str, pos);
  const out = [];
  for (const c of cps) {
    if (c < 0x80) {
      out.push(c);
    } else if (c < 0x800) {
      out.push(0xc0 | (c >> 6), 0x80 | (c & 0x3f));
    } else if (c < 0x10000) {
      out.push(0xe0 | (c >> 12), 0x80 | ((c >> 6) & 0x3f), 0x80 | (c & 0x3f));
    } else {
      out.push(
        0xf0 | (c >> 18),
        0x80 | ((c >> 12) & 0x3f),
        0x80 | ((c >> 6) & 0x3f),
        0x80 | (c & 0x3f),
      );
    }
  }
  return Uint8Array.from(out);
}

// Strict: rejects overlong forms, surrogates, values above U+10FFFF and
// truncated sequences. No replacement characters, ever.
export function decodeUtf8(bytes, pos) {
  const cps = [];
  const n = bytes.length;
  let i = 0;
  while (i < n) {
    const b = bytes[i];
    let need, cp, lo, hi;
    if (b < 0x80) {
      cps.push(b);
      i++;
      continue;
    } else if (b >= 0xc2 && b <= 0xdf) {
      need = 1; cp = b & 0x1f; lo = 0x80; hi = 0xbf;
    } else if (b === 0xe0) {
      need = 2; cp = 0; lo = 0xa0; hi = 0xbf;      // reject overlong 3-byte
    } else if (b >= 0xe1 && b <= 0xec) {
      need = 2; cp = b & 0x0f; lo = 0x80; hi = 0xbf;
    } else if (b === 0xed) {
      need = 2; cp = 0x0d; lo = 0x80; hi = 0x9f;   // reject surrogates
    } else if (b >= 0xee && b <= 0xef) {
      need = 2; cp = b & 0x0f; lo = 0x80; hi = 0xbf;
    } else if (b === 0xf0) {
      need = 3; cp = 0; lo = 0x90; hi = 0xbf;      // reject overlong 4-byte
    } else if (b >= 0xf1 && b <= 0xf3) {
      need = 3; cp = b & 0x07; lo = 0x80; hi = 0xbf;
    } else if (b === 0xf4) {
      need = 3; cp = 4; lo = 0x80; hi = 0x8f;      // cap at U+10FFFF
    } else {
      fail('E_UTF8', `invalid start byte 0x${b.toString(16)} at byte ${i}`, pos);
    }

    if (i + need >= n) {
      fail('E_UTF8', `truncated sequence at byte ${i}`, pos);
    }
    for (let k = 1; k <= need; k++) {
      const c = bytes[i + k];
      const min = k === 1 ? lo : 0x80;
      const max = k === 1 ? hi : 0xbf;
      if (c < min || c > max) {
        fail('E_UTF8', `invalid continuation byte at byte ${i + k}`, pos);
      }
      cp = (cp << 6) | (c & 0x3f);
    }
    cps.push(cp);
    i += need + 1;
  }
  return fromCodePoints(cps);
}

export function bytesToHex(bytes) {
  let out = '';
  for (let i = 0; i < bytes.length; i++) {
    out += bytes[i].toString(16).padStart(2, '0');
  }
  return out;
}

export function bytesEqual(a, b) {
  if (a.length !== b.length) return false;
  for (let i = 0; i < a.length; i++) if (a[i] !== b[i]) return false;
  return true;
}

// Bytewise, as the spec requires — JS's native comparison is UTF-16 order and
// disagrees above U+FFFF.
export function bytesCompare(a, b) {
  const n = Math.min(a.length, b.length);
  for (let i = 0; i < n; i++) {
    if (a[i] !== b[i]) return a[i] < b[i] ? -1 : 1;
  }
  return a.length === b.length ? 0 : (a.length < b.length ? -1 : 1);
}

File: python/sel/utf8.py
"""UTF-8 codec, hand-written on purpose.

The host's own facilities are not used. `bytes.decode('utf-8')` raises
UnicodeDecodeError rather than E_UTF8 and its error positions are byte offsets
into a message string; `str.encode` refuses lone surrogates with a different
exception again. Both would leak a host exception type through an API whose
whole contract is that every failure is a SelError.

Python does give one thing free that the other hosts pay for: `str` is a
sequence of code points, so `len()` and slicing already count what SEL counts.
That is why `to_code_points` here is nearly the identity — it exists to *reject*
lone surrogates, which Python permits in a `str` and UTF-8 has no encoding for.
"""

from __future__ import annotations

from .errors import Pos, fail


def to_code_points(s: str, pos: Pos | None = None) -> list[int]:
    """Code points, rejecting the lone surrogates Python allows in a str.

    `chr(0xD800)` is a legal Python str element and has no UTF-8 encoding, so it
    cannot be a SEL TEXT value. The other hosts hit the same wall from the other
    side: JS strings are UTF-16 and produce lone surrogates by slicing.
    """
    out = []
    for ch in s:
        c = ord(ch)
        if 0xD800 <= c <= 0xDFFF:
            which = 'high' if c <= 0xDBFF else 'low'
            fail('E_UTF8', f'unpaired {which} surrogate', pos)
        out.append(c)
    return out


def from_code_points(cps: list[int]) -> str:
    return ''.join(map(chr, cps))


def encode_utf8(s: str, pos: Pos | None = None) -> bytes:
    cps = to_code_points(s, pos)
    out = bytearray()
    for c in cps:
        if c < 0x80:
            out.append(c)
        elif c < 0x800:
            out += bytes((0xC0 | (c >> 6), 0x80 | (c & 0x3F)))
        elif c < 0x10000:
            out += bytes((0xE0 | (c >> 12), 0x80 | ((c >> 6) & 0x3F), 0x80 | (c & 0x3F)))
        else:
            out += bytes((0xF0 | (c >> 18), 0x80 | ((c >> 12) & 0x3F),
                          0x80 | ((c >> 6) & 0x3F), 0x80 | (c & 0x3F)))
    return bytes(out)


def decode_utf8(data: bytes, pos: Pos | None = None) -> str:
    """Strict: rejects overlong forms, surrogates, values above U+10FFFF and
    truncated sequences. No replacement characters, ever.
    """
    cps: list[int] = []
    n = len(data)
    i = 0
    while i < n:
        b = data[i]
        if b < 0x80:
            cps.append(b)
            i += 1
            continue
        if 0xC2 <= b <= 0xDF:
            need, cp, lo, hi = 1, b & 0x1F, 0x80, 0xBF
        elif b == 0xE0:
            need, cp, lo, hi = 2, 0, 0xA0, 0xBF          # reject overlong 3-byte
        elif 0xE1 <= b <= 0xEC:
            need, cp, lo, hi = 2, b & 0x0F, 0x80, 0xBF
        elif b == 0xED:
            need, cp, lo, hi = 2, 0x0D, 0x80, 0x9F       # reject surrogates
        elif 0xEE <= b <= 0xEF:
            need, cp, lo, hi = 2, b & 0x0F, 0x80, 0xBF
        elif b == 0xF0:
            need, cp, lo, hi = 3, 0, 0x90, 0xBF          # reject overlong 4-byte
        elif 0xF1 <= b <= 0xF3:
            need, cp, lo, hi = 3, b & 0x07, 0x80, 0xBF
        elif b == 0xF4:
            need, cp, lo, hi = 3, 4, 0x80, 0x8F          # cap at U+10FFFF
        else:
            fail('E_UTF8', f'invalid start byte 0x{b:x} at byte {i}', pos)

        if i + need >= n:
            fail('E_UTF8', f'truncated sequence at byte {i}', pos)
        for k in range(1, need + 1):
            c = data[i + k]
            lo_k = lo if k == 1 else 0x80
            hi_k = hi if k == 1 else 0xBF
            if c < lo_k or c > hi_k:
                fail('E_UTF8', f'invalid continuation byte at byte {i + k}', pos)
            cp = (cp << 6) | (c & 0x3F)
        cps.append(cp)
        i += need + 1
    return from_code_points(cps)


def bytes_to_hex(data: bytes) -> str:
    return data.hex()


def bytes_equal(a: bytes, b: bytes) -> bool:
    return a == b


def bytes_compare(a: bytes, b: bytes) -> int:
    """Bytewise, as the spec requires.

    Python's own `<` on bytes is already bytewise and would give the same answer,
    and on str it compares code points, which happens to agree with UTF-8 byte
    order too. Neither shortcut is taken: the spec says bytes, the other three
    hosts compare bytes explicitly because their shortcuts are *wrong*, and the
    next reader should not have to work out that Python's are safe.
    """
    n = min(len(a), len(b))
    for i in range(n):
        if a[i] != b[i]:
            return -1 if a[i] < b[i] else 1
    if len(a) == len(b):
        return 0
    return -1 if len(a) < len(b) else 1
